Mobile Buy SDK/Mobile Buy SDK/Data/BUYClient.m

```diff
@@ -424,75 +419,33 @@ - (NSURLSessionDataTask *)getCheckout:(BUYCheckout *)checkout completion:(BUYDat
-			[self handleCheckoutResponse:json error:error block:block];
-		}];
-	}
-	return task;
+	NSURLComponents *components = [self URLComponentsForCheckoutsAppendingPath:nil
+																 checkoutToken:checkout.token
+																	queryItems:nil];
+	return [self patchRequestForURL:components.URL body:data completionHandler:^(NSDictionary *json, NSURLResponse *response, NSError *error) {
+		[self handleCheckoutResponse:json error:error block:block];
+	}];
 }
 
-- (NSURLSessionDataTask*)completeCheckout:(BUYCheckout *)checkout completion:(BUYDataCheckoutBlock)block
+- (NSURLSessionDataTask*)completeCheckout:(BUYCheckout *)checkout paymentToken:(id<BUYPaymentToken>)paymentToken completion:(BUYDataCheckoutBlock)block
 {
-	NSURLSessionDataTask *task = nil;
+	BUYAssertCheckout(checkout);
 	
-	if ([checkout hasToken]) {
-		
-		NSData *data = nil;
-		NSError *error = nil;
-		
-		if (checkout.paymentSessionId.length > 0) {
-			NSDictionary *paymentJson = @{ @"payment_session_id" : checkout.paymentSessionId };
-			data = [NSJSONSerialization dataWithJSONObject:paymentJson options:0 error:&error];
-		}
-		
-		if ((data && !error) || (checkout.paymentDue && checkout.paymentDue.floatValue == 0)) {
-			task = [self checkoutCompletionRequestWithCheckout:checkout body:data completion:block];
-		}
-	}
-	else {
-		block(nil, [NSError errorWithDomain:kShopifyError code:BUYShopifyError_InvalidCheckoutObject userInfo:nil]);
+	NSData *data = nil;
+	if (paymentToken) {
+		data = [NSJSONSerialization dataWithJSONObject:[paymentToken JSONDictionary] options:0 error:nil];
 	}
 	
-	return task;
-}
-
-- (NSURLSessionDataTask *)completeCheckout:(BUYCheckout *)checkout withApplePayToken:(PKPaymentToken *)token completion:(BUYDataCheckoutBlock)block
-{
-	
-	NSURLSessionDataTask *task = nil;
-#if __has_include(<PassKit/PassKit.h>)
+	BOOL isFree = (checkout.paymentDue && checkout.paymentDue.floatValue == 0);
 	
-	if ([checkout hasToken] == NO) {
-		block(nil, [NSError errorWithDomain:kShopifyError code:BUYShopifyError_InvalidCheckoutObject userInfo:nil]);
-	}
-	else if (!token) {
-		block(nil, [NSError errorWithDomain:kShopifyError code:BUYShopifyError_NoApplePayTokenSpecified userInfo:nil]);
-	}
-	else {
-		NSString *tokenString = [[NSString alloc] initWithData:token.paymentData encoding:NSUTF8StringEncoding];
-		NSDictionary *paymentJson = @{ @"payment_token" : @{ @"payment_data" : tokenString, @"type" : @"apple_pay" }};
-		NSError *error = nil;
-		NSData *data = [NSJSONSerialization dataWithJSONObject:paymentJson options:0 error:&error];
-		if (data && !error) {
-			task = [self checkoutCompletionRequestWithCheckout:checkout body:data completion:block];
-		}
-		else {
-			block(nil, [NSError errorWithDomain:kShopifyError code:BUYShopifyError_InvalidCheckoutObject userInfo:nil]);
-		}
-	}
+	NSAssert(data || isFree, @"Failed to complete checkout. Checkout must have a payment token or have a payment value equal to $0.00");
```

Comment:
I prefer that all asserts on arguments are at the very top of the method, for easier reading. In this case, `data` depends on `paymentToken`, so we should just assert (or throw invalid arguments exception) that `paymentToken` is valid. Could add a read-only `valid (getter=isValid)` property to `BUYPaymentToken`.